File: ProjectCS/UI/StudentPage.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SessionResultsCS.UI
{
    public class StudentPage : LoopPage
    {
        private Student _student;

        private MenuActionPicker _picker;


        public StudentPage(Student student)
        {
            _student = student;
            _picker = new MenuActionPicker([
                ("Изменить номер студенческого билета", ChangeStudentID),
                ("Изменить имя студента", ChangeStudentFirstName),
                ("Изменить фамилию студента", ChangeStudentLastName),
                ("Изменить отчество студента", ChangeStudentMiddleName),
                ("Назад", StopLoop),
            ]);
        }

        protected override void Loop()
        {
            Console.WriteLine($"== {_student} ==");
            Console.WriteLine("Действия:");
            _picker.PickAndRun();
        }


        private void ChangeStudentID()
        {
            try
            {
                Console.Write("Введите новый номер студенческого билета: ");
                string studentID = Console.ReadLine()!;
                _student.StudentID = uint.Parse(studentID);
            }
            catch (Exception e) when (e is ArgumentException || e is FormatException)
            {
                Console.WriteLine("Введен некорректный курс");
            }
        }

        private void ChangeStudentFirstName()
        {
            try
            {
                Console.Write("Введите новое имя студента: ");
                string firstName = Console.ReadLine()!;
                _student.FirstName = firstName;
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Введено некорректное имя");
            }
        }

        private void ChangeStudentLastName()
        {
            try
            {
                Console.Write("Введите новую фамилию студента: ");
                string lastName = Console.ReadLine()!;
                _student.LastName = lastName;
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Введена некорректная фамилия");
            }
        }

        private void ChangeStudentMiddleName()
        {
            try
            {
                Console.Write("Введите новое отчество студента: ");
                string middleName = Console.ReadLine()!;
                _student.MiddleName = middleName;
            }
            catch (ArgumentException)
            {
                Console.WriteLine("Введено некорректное отчество");
            }
        }
    }
}

File: Project/UI/StudentPage.cpp
#include "UI/StudentPage.hpp"

#include <format>
#include <iostream>


StudentPage::StudentPage(std::shared_ptr<Student> student) {
	m_student = student;
	SetPicker(std::make_unique<MenuActionPicker>(std::vector<MenuActionPicker::Item>{
		{ "Изменить номер студенческого билета", [this]() { ChangeStudentID(); } },
		{ "Изменить имя студента", [this]() { ChangeStudentFirstName(); } },
		{ "Изменить фамилию студента", [this]() { ChangeStudentLastName(); } },
		{ "Изменить отчество студента", [this]() { ChangeStudentMiddleName(); } },
		{ "Назад", [this]() { StopLoop(); } }
	}));
}

void StudentPage::Loop() {
	std::cout << std::format("== {} ==\n", m_student->ToString());
}

void StudentPage::ChangeStudentID() {
	try {
		std::cout << "Введите новый номер студенческого билета: ";
		std::string studentIDInput;
		std::getline(std::cin, studentIDInput);
		unsigned int studentID = std::stoul(studentIDInput);
		m_student->SetStudentID(studentID);
	} catch (std::invalid_argument) {
		std::cout << "Введен некорректный номер\n";
	} catch (std::out_of_range) {
		std::cout << "Введен некорректный номер\n";
	}
}

void StudentPage::ChangeStudentFirstName() {
	try {
		std::cout << "Введите новое имя студента: ";
		std::string firstName;
		std::getline(std::cin, firstName);
		m_student->SetFirstName(firstName);
	} catch (std::invalid_argument) {
		std::cout << "Введено некорректное имя\n";
	}
}

void StudentPage::ChangeStudentLastName() {
	try {
		std::cout << "Введите новую фамилию студента: ";
		std::string lastName;
		std::getline(std::cin, lastName);
		m_student->SetLastName(lastName);
	} catch (std::invalid_argument) {
		std::cout << "Введена некорректная фамилия\n";
	}
}

void StudentPage::ChangeStudentMiddleName() {
	try {
		std::cout << "Введите новое отчество студента: ";
		std::string middleName;
		std::getline(std::cin, middleName);
		m_student->SetMiddleName(middleName);
	} catch (std::invalid_argument) {
		std::cout << "Введено некорректное отчество\n";
	}
}